src/room.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017 Tunnler Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "tunnler/room.h"
+
+#include "tunnler/room_message_types.h"
+
+#include "BitStream.h"
+
+/// Maximum number of concurrent connections allowed to this room.
+static const uint32_t MaxConcurrentConnections = 10;
+
+void Room::Create(const std::string& name, const std::string& server_address, uint16_t server_port) {
+    RakNet::SocketDescriptor socket(server_port, server_address.c_str());
+
+    server = RakNet::RakPeerInterface::GetInstance();
+    // TODO(Subv): Allow specifying the maximum number of concurrent connections.
+    server->Startup(MaxConcurrentConnections, &socket, 1);
+    server->SetMaximumIncomingConnections(MaxConcurrentConnections);
+
+    state = State::Open;
+
+    // Start a network thread to Receive packets in a loop.
+    room_thread = std::make_unique<std::thread>(&Room::ServerLoop, this);
+}
+
+void Room::Destroy() {
+    state = State::Closed;
+
+    {
+        std::lock_guard<std::mutex> lock(server_mutex);
+
+        if (server) {
+            server->Shutdown(300);
+            RakNet::RakPeerInterface::DestroyInstance(server);
+        }
+        server = nullptr;
+    }
+
+    room_thread->join();
```

Comment:
the room_thread is a `unique_ptr` / doesn't the thread get destroyed automaticly?
This is only do to things gracefully - we already assert in the other case so API users should know that shit is about to go down (we should just document the fact that rooms must be `destroy()`d before destroying the object).
However, maybe we *must* forcefully [not the same as `leave()`!] kill the thread to avoid race conditions where the room_thread will use an already closed / destroyed RakPeer - but we might be doing that already, I didn't check